Add support for RFC 8707 OAuth2 Resource Indicators (#14355)

thomasdarimont opened the pull request:
Add support for [RFC 8707 OAuth2 Resource Indicators](https://www.rfc-editor.org/rfc/rfc8707) (#14355)

Discussion: https://github.com/keycloak/keycloak/discussions/35743

- Adds `ResourceIndicatorMapper` OIDC protocol mapper to manage the `aud` claim based on the `resource` parameter
- Make `AuthorizationEndpoint` and `TokenEndpoint` aware of `resource` parameter.
- Enhanced Authz request parsing to support repeated parameter (resource)
- Added `OAuth2ResourceIndicatorResolver` SPI to allow filtering of resource indicators and lookup associated clients for resource indicators

If the authorization request contains one or more resource parameters, those resource values are first checked against the list of allowed resource indicators for the given client. Allowed resource indicators are determined
via the `OAuth2ResourceIndicatorResolver` SPI.
If the resource indicators are legit for the client, the initially requested resource identifier set is stored in the client session. 
Token Requests / Token Refresh Requests can later request access tokens minted to support resource indicators in the `aud` claim based on the resource indicators that were initially requested.

Note that this allows users to dynamically extend or narrow down the `aud` claim based on the given resource identifiers in token / token refresh requests.

Fixes #14355

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
Simple usage example

# Client config for myclient

ResourceIndicatorMapper for myclient cotains the allow resource identifier patterns:
- foo
- bar
- baz

# Initial Auth Request:
```
$ISSUER/protocol/openid-connect/auth
?client_id=myclient
&redirect_uri=...
&state=...
&response_mode=fragment
&response_type=code
&scope=openid
&nonce=...
&code_challenge=...
&code_challenge_method=S256
&resource=foo
&resource=bar
```

-> This restricts the current auth session to use only `foo` and `bar` as resource indicators, even if the client myclient would also allow `baz`.


# Token Request
```
POST $ISSUER/protocol/openid-connect/token
grant_type=authorization_code
&client_id=myclient
&code=...
&resource=foo
...
```

Return:
Access Token with audience:
```json
{
    ...
    aud: [foo]
}
```

# Token Refresh Request
```
POST $ISSUER/protocol/openid-connect/token
Authorization: ...
Content-Type: application/x-www-form-urlencoded
grant_type=refresh_token
&refresh_token=...
&resource=bar
...
```

Return:
Access Token with audience:
```json
{
    ...
    aud: [bar]
}
```

stianst commented:
Haven't had a chance to look at this properly or think too much about it; but feels a bit like shoehorning/abusing protocol mappers, and not a natural way to manage resources for a client.

stianst commented:
Few random thoughts around this:

* We'll probably eventually need some way of querying clients given a resource indicator
* Resource indicators according to the spec must be a URI
* RFC8693 has audience and resource has optional parameters; that means it's perfectly valid to only specify the resource parts, which again means there needs to be a efficient way to lookup the client for a given resource
* Clients can have a base-url, kinda makes sense to allow configuring resources relative to that, rather than having to specify the full URI, at least in most common cases
* Not sure how this all fits together with authz services, but there's definitively some overlap between resources defined for authz and resource indicators

thomasdarimont commented:
Hi @stianst, thanks for your helpful comments! I created the following discussion to revise the design of the resource identifiers support: https://github.com/keycloak/keycloak/discussions/35743

In this discussion I address the concerns you mentioned.

thomasdarimont commented:
Yes, I agree this is somewhat abusing the protocol mappers mechanism beyond token generation since it also alters the Authorization Server behaviour. However, I felt quite pragmatic to do so. However, I agree that this protocol mapper usage might be uncommon - but I have to start somewhere :O) 

Also the implementation can easily changed to load the resource identifiers list from somewhere else :)

> We'll probably eventually need some way of querying clients given a resource indicator
I understand the use-case. I added the point to the discussion.

> Resource indicators according to the spec must be a URI
Yes, that's written in the spec and the resource examples I gave were just that, examples. Quite often Keycloak users also model resource servers / APIs as clients without flows in Keycloak to configure client specific roles etc, that can be used by other clients. In my mind resource identifiers == client_id for resource server clients, my be this is a bit off, but since client_ids are usually not URIs (at least it's not enforced) I used non-URI values as resources in the examples.
I'm fine with using URIs though which would probably support your requirement for performing lookups for clients by resource identifier.

> RFC8693 has audience and resource has optional parameters; that means it's perfectly valid to only specify the resource parts, which again means there needs to be a efficient way to lookup the client for a given resource
Noted, and added to the discussion.

> Clients can have a base-url, kinda makes sense to allow configuring resources relative to that, rather than having to specify the full URI, at least in most common cases
Sometimes the configured base URL might not necessarily match the resource identifiers. I'm thinking of deployments of a resource server in different environments (dev, test, qa, prod), where the base URL is different but the logical resource URIs might be the same for all environments.

> Not sure how this all fits together with authz services, but there's definitively some overlap between resources defined for authz and resource indicators
I think the resources in authz services could indeed be used for this, but this would lead to a stronger coupling of the authorization services with the oauth2 functionality which might cause other problems.
Thinking about this a bit, we could use authorization resources to configure the allowed resource identifiers for a client.
This can already be implemented quite easily in this PR: I just have to adjust the new [org.keycloak.protocol.oauth2.ResourceIndicators#getSupportedResourceIndicators(..)](https://github.com/keycloak/keycloak/pull/35711/files#diff-fdc1ce99a8caed471d20f1b04c094a623e3da887710a595e75acecf199213d51R102) method to fetch the resource URIs from authorization services instead of the protocol mapper configuration.

We then would also need a way to let clients "opt-in" to the resource identifier support somehow, perhaps with an "advanced setting" in the client configuration: `Resource Identifier Support Enabled: on/off` default off.

thomasdarimont commented:
I refactored this the resource indicator support a bit:

- Introduced `OAuth2ResourceIndicatorsProvider` SPI to allow users to specify their own ways for validating resource identifiers (see O4) Add new dedicated SPI to validate allowed resource indicators R1 in https://github.com/keycloak/keycloak/discussions/35743)
- Removed the ability to configure supported resource indicators from the Resource Indicator Mapper

We could then add a default implementation that checks given resource identifiers based on something like a component config (config per client) etc.

thomasdarimont commented:
I updated the PR with the a way to configure client specific resource indicators with the help of the authorization services:

![image](https://github.com/user-attachments/assets/fec9f93c-0d6a-472f-84eb-67bb609cad95)
User can now specify custom resources with the special type URI: `urn:keycloak:resource-indicator` to denote a resource as resource indicator.

To apply the resource indicator semantics to the aud claim of a token, users need to configure a dedicated "resource-indicator-mapper", 
![image](https://github.com/user-attachments/assets/8d93527e-82b9-43a1-8ed5-af3c131b3e5e)

cgeorgilakis commented:
Nice work 👏.
What happen during token exchange OIDC to OIDC client in aud claim if mapper exists and resource parameter exists as request parameter?

thomasdarimont commented:
Thanks @cgeorgilakis for your feedback. The handling of the `resource` parameter is out of scope for this PR, as the resource handling for token exchange is covered by the [Token Exchange rfc8693](https://datatracker.ietf.org/doc/html/rfc8693#section-2.1) and should be covered by separate PR, e.g. https://github.com/keycloak/keycloak/issues/31553 or https://github.com/keycloak/keycloak/pull/36848.

thomasdarimont commented:
Client photoz client
<img width="806" alt="image" src="https://github.com/user-attachments/assets/9790ad9c-4119-4c63-bfbc-ad636ab4a07a" />

... with authorization enabled
<img width="335" alt="image" src="https://github.com/user-attachments/assets/8588d673-d25b-4404-aa6d-d070d8c737ed" />

Resources can be defined via Authorization / Resources in the client:
<img width="1017" alt="image" src="https://github.com/user-attachments/assets/9844c73f-ea40-4680-a182-56d1d8d2deeb" />

The custom type: `urn:keycloak:oauth2:resource-indicator` declares a resource indicator, the URI specifies the actual URI: `https://api.acme.com/photos`

Available resource indicators for a client can be seenin the resources tab:
<img width="1244" alt="image" src="https://github.com/user-attachments/assets/5b30c4ba-a353-4e00-81cc-5212bd104d35" />

A new `Resource Indicators` protocol mapper exposes the configured resources in the `aud` claim in addition to the existing linked clientIds.
<img width="824" alt="image" src="https://github.com/user-attachments/assets/0244db46-2e51-4ca6-a0c5-d0aa6ab5ef0e" />

Dedicated Protocol Mapper:
<img width="1233" alt="image" src="https://github.com/user-attachments/assets/2d73e877-78d7-45dd-ae4a-00523a8bc986" />

One can find an example realm and some http request examples with curl here: https://gist.github.com/thomasdarimont/0ccab1238d3824a84a37a16a086cf890

thomasdarimont committed:
Add support for RFC 8707 OAuth2 Resource Indicators (#14355)

- Adds ResourceIndicatorMapper OIDC protocol mapper
- Make AuthorizationEndpoint and Token Endpoint aware of resource parameter.
- Enhanced Authz request parsing to support repeated parameter (resource)

If the authorization request contains one or more resource parameters, those resource values are first checked against
the list of allowed resource indicator patterns for the given client.
If the resource indicators are legit for the client, the initially requested resource identifier set is stored in the client session.
Token Requests / Token Refresh Requests can later request access tokens minted to support resource indicators in the aud claim based on the resource indicators that were initially requested.

Note that this allows users to dynamically extend or narrow down
the audience based on the given resource identifiers in token / token refresh requests.

The allowed resource identifiers for a client are managed via the ResourceIndicatorMapper.
ResourceIndicatorMapper manages aud claim via resource parameters. Users can configure the list of allowed resource indicator patterns in the ResourceIndicatorMapper configuration for a client. Note that only resource indicators allowed by the mapper configuration can be requested for a client.

Fixes #14355

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Refactor ResourceIndicator handling

- Introduced OAuth2ResourceIndicatorsProvider SPI to abstract how valid Resource Indicators for clients are determined
- Revised the resource parameter checking
- Removed configuration of resource indicators from ResourceIndicatorMapper

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Add an example default implementation for OAuth2ResourceIndicatorsProvider

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Revise code in flattenDecodedFormParametersToParamsMap

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Revise OAuth2ResourceIndicatorsProvider

- Rename SPI to resource-indicator-resolver
- Rename to OAuth2ResourceIndicatorsProvider
- Add ability to lookup clients by resource indicators

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Next iteration for RFC 8707 OAuth2 Resource Indicators support

- We now allow specifying supported resource indicators via the authorizations tab in the client configuration.
- Users can declare new resources with the special type URI: `urn:keycloak:resource-identifier` to denote an resource identifier.
- According to the RFC 8707, resource indicators must be URIs, but we also support Keycloak clientIds as allowed resource indicators to support existing Keycloak realm configurations.

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Change KEYCLOAK_RESOURCE_INDICATOR_TYPE to urn:keycloak:oauth2:resource-identifier

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Revise DefaultOAuth2ResourceIndicatorResolver

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Validate resource parameter during token exchange.

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

Change KEYCLOAK_RESOURCE_INDICATOR_TYPE to urn:keycloak:oauth2:resource-indicator

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

# Conflicts:
#	services/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantTypeBase.java
#	services/src/main/java/org/keycloak/protocol/oidc/tokenexchange/AbstractTokenExchangeProvider.java

stianst reviewed (changes requested):
We should limit the initial support of resource indicators to a client having a single resource uri.

That brings the question of what resource can a client requests; for that we would need a way to link the client to the audience clients. Which is something we are aiming to deliver in #43179

We could leverage client roles though for this purpose perhaps? We could use something like the following:

1. Add an optional Resource URI attribute to clients
2. Get a list of all clients the client has a client role scope on
3. For each `resource` parameter check it matches the `Resource URI` of one of the clients from 2

I changed this to a draft since there are no tests at all; I'd suggest we break this up into multiple stages starting with only supporting `resource` parameter in authorization request and token requests grant types `authorization_code` and `refresh_token`.

thomasdarimont commented:
Thanks @stianst for picking this up again. I don't have time to work on this today / next few days. So if you want to work on this now feel free to adopt it.

Moving forward, instead of continuing this PR, we should close it and create a new one.
I think we can salvage some of the ideas from this PR and add this feature in a stepwise manner.

Some questions/thoughts regarding your proposal:

> We could use something like the following:
> 
> 1. Add an optional Resource URI attribute to clients
> 2. Get a list of all clients the client has a client role scope on
> 3. For each resource parameter check it matches the Resource URI of one of the clients from 2

Q1) How to configure the (primary) resource indicator for a client?
For 1. Would you like to use a new client attribute or a dedicated property for this?
I initially thought of a client attribute, but the (primary) resource indicator must be unique among all clients in a realm. I'm not sure if we already do this somewhere, but we may need to add a new field to the client JPA entity to store the resource indicator URI, making it easier to ensure uniqueness at the database level. 
We can then expose this as a dedicated client property (more work, which might require admin API consumer updates) or pass it through the attributes (a hacky solution, but it would not break admin API consumers).

With this, we would only allow one (primary?) resource indicator per client for now, which can be used as an alias to target this specific client/resource server via the `resource` parameter.
However, later we may need to support additional (sub-resource) indicators per resource server (e.g., if those clients represent an API gateway or another collection of resource servers). Those additional resource indicators might come from another source, but we can add support for that later. In my PR, I introduced an OAuth2ResourceIndicatorResolver SPI for this.

Q2) Mechanism to select the eligible resource indicators for a given client
Do you mean to use the client role scope for selecting other related clients and derive the resource indicator from that?
Doesn't this require clients to specify at least one client role for selection?

To verify this:
Imagine we have a `photos-app` OIDC client that can obtain tokens and two APIs that the `photos-app` should be able to access: `photos-api`, `accounts-api`.
We create `photos-api` and `accounts-api` clients as a resource server (all flows disabled). For both, we specify resource indicators, such as `https://accounts.acme.test` and `https://photos.acme.test`.
We now want to express that a `photos-app` token should be able to access `photoz-api` and `accounts-api` via the `resource` parameter in the authorization / token request. With your method, I need to define at least one role in the `photoz-api` and `accounts-api` clients to be able to add them to the client role scope selection of the `photoz-app`. So I create a "dummy" role like `access` for `photos-api`, `accounts-api`, and add them to the client role scope of the `photos-app`.

Is this what you had in mind?

Q3) How to match a client via resource indicator to the requested `resource` parameter?

Using the client role scope to select clients and derive the eligible resource indicators is a simple solution. Still, it requires additional configuration (role definitions) and is also limited to just one indicator (which is fine for now). However, I wonder if we should make this pluggable from the start and use a new SPI to derive eligible resource indicators for a client with it (a default implementation could use the client role scope for this purpose).
I believe the clients selected need to have a resource indicator URI configured to be considered for matching the `resource` parameter. We could filter out clients without a resource indicator.

Q4) How do we specify whether a client supports the resource parameter or not?
Is this enabled for all clients automatically? In my PR, I added a dedicated `ResourceIndicatorMapper` OIDC protocol mapper that admins could add to their clients to explicitly enable the handling of `resource` parameters, which added eligible resource indicator values to the `aud` claim after some filtering.

Q5) When and where do we determine the eligible resource indicators?
In my PR, the set of eligible resource indicators was determined during authorization request handling and subsequently filtered based on the actual token request, allowing users to request a subset of the initially requested resources.

thomasdarimont commented:
Yes let's do this together :) I can send a new PR with just the minimal bits to support a new "Root Resource URI" client attribute for clients. I guess this setting should be placed in the "Advanced Settings" somehow and be empty by default.
